EspressoSystems/espresso-sequencer
# Add Inscriptions Service

### This PR:
Tracks the overall progress of the implementation of the Inscriptions service.  It is specifically targeting a merge into `release-rogue` as that is what it was branched from.  The `release-rogue` branch was targeted explicitly to ensure compatability.  This is a DRAFT PR to ensure that it doesn't get directly merged.

This adds the `inscriptions` sub-crate into the sequencer repo for convenience of writing.  It is largely copied from the `node-metrics` implementation with state and message adjustments to work specifically for the inscriptions API itself.

This service is meant to be used with the inscriptions front-end ui demo.  Much of the implementation has been implemented quickly without a lot of time for testing or documentation due to time constraints.

### How to Test:

This service has a few environment variables that are important to be specified in order for things to be able to be run without issue.  Some of these environment variables should have special care taken when specifying them in order to ensure that the behavior can be run and verified in a reasonable time frame.

#### Environment Variables

The environment variables for this service are utilized to configure specific behavior, generally around targetting specific entries and data sources.

##### Persistence

This service utilizes `postgres` and `sqlx` to store all of its state.  It is important to have a postgres instance running, and configured properly in order for this service to run.

Before configuring the environment variables to run the service, the database needs to be setup and configured.  With a properly running `postgres` issue it is important to initialize the `postgres` database utilizing the `sqlx-cli`:

https://github.com/launchbadge/sqlx/blob/main/sqlx-cli/README.md

the `DATABASE_URL` environment variable **NEEDS** to be set for the `sqlx-cli` to be able to run.  After that, ensure that you run this command, with the environment variable setup correctly:

```sh
sqlx database create
```

Then these environment variables can be set in order to run migrations and make things work:
- POSTGRES_HOST
- POSTGRES_PORT
- POSTGRES_DATABASE
- POSTGRES_USER
- POSTGRES_PASSWORD

These environment variables have default values with the idea that the postgres server is accessible via `localhost` as if running in a `docker` container running locally. 

##### Sourcing Data

This service sources all of its data from the availability API's block stream endpoint.  This endpoint needs to be specified:
- ESPRESSO_INSCRIPTIONS_BLOCK_STREAM_SOURCE_BASE_URL

This expected value is meant to be the base version endpoint.

Example:
`https://query.decaf.testnet.espresso.network/v0/`

Additionally the variable `ESPRESSO_INSCRIPTIONS_MINIMUM_BLOCK_HEIGHT` **SHOULD** be specified in order to ensure that we don't start consuming all blocks from block height `0` (which is the default).

- ESPRESSO_INSCRIPTIONS_MINIMUM_BLOCK_HEIGHT

The first block on `decaf` that has an inscription with the default namespace id is `453728`.

The data coming in will be scanned for transactions that match the namespace and appear to be signed by the private key of the service.  Otherwise they get ignored.

The key for the service can be setup with a BIP39 mnemonic:
- ESPRESSO_INSCRIPTIONS_SIGNER_MNEMONIC

The namespaceID can be changed with:
- ESPRESSO_INSCRIPTIONS_NAMESPACE_ID


##### Submitting Data

The data will be submitted to a mempool of some sort. It is necessary to specify a URL endpoint to hit for submitting data.  This endpoint matches the configuration of the sequencer service.  You can target either public mempool submit endpoints or a private (the builder's) endpoint.

- ESPRESSO_INSCRIPTIONS_SUBMIT_BASE_URL

Examples:
`https://query.decaf.testnet.espresso.network/v0/submit`
`https://builder.decaf.testnet.espresso.network/v0/txn_submit`

Submissions get rate limited, and buffered to a configurable rate and buffer size using these environment variables:

- ESPRESSO_INSCRIPTIONS_PUT_INSCRIPTION_BUFFER_SIZE
- ESPRESSO_INSCRIPTIONS_PUT_INSCRIPTIONS_PER_SECOND
